### PR TITLE
Implement powersink surge

### PR DESCRIPTION
**Pull request by skull132**
Implements suggestion https://forums.aurorastation.org/viewtopic.php?f=21&p=78011 in the following fashion:
- Powersinks explode ~18 minutes after being placed on normal SMES setup. Obviously more input from engineering will make this process go faster.
-  Upon reaching their capacity, they will now cause a larger area power surge. The surge will smash all lights belonging to APCs within close to moderate proximity, and call EMP act on all connected powernet items that are within range (severity depending on distance). All of those items have a small chance to cause a minor explosion as well, primarily because EMP act is fucking whimpy.

[Preview](https://kama.skullnet.me/index.php/s/AE6ejFJghToXjmG)

Note that since that preview was done, the chance of exploding has been lowered, so there won't be as many.

### CODE REVIEW

File: code/_helpers/maths.dm [hunk 1]
```diff
@@ -145,3 +145,7 @@
     . = num
     for (var/i = 0, i < iterations, i++)
         . = (1/3) * (num/(.**2)+2*.)
+
+// A neat little helper to convert the value X, that's between imin and imax, into a value
+// that's proportionally scaled and in range of omin and omax.
+#define MAP(x, imin, imax, omin, omax) ((x - imin) * (omax - omin) / (imax - imin) + omin)
```
Comment (Lohikar) [inline code/_helpers/maths.dm]: This should probably be in `__defines`, not `_helpers`.

File: code/game/objects/items/devices/powersink.dm [hunk 2]
```diff
@@ -120,13 +121,62 @@
...
+		// Check for terminals and affect their master nodes. Also add a special
+		// case for APCs whereby their lights are popped or flicked.
+		if (istype(A, /obj/machinery/power/terminal))
+			var/obj/machinery/power/terminal/T = A
+			if (istype(T.master, /obj/machinery/power/apc))
+				var/obj/machinery/power/apc/AP = T.master
+				if (dist > 1)
+					AP.overload_lighting(100, TRUE)
+				else
+					AP.flicker_all()
```
Comment (Lohikar) [inline code/game/objects/items/devices/powersink.dm]: How many timers do you see when this is triggered? I recall this proc creating ~100 timers for just a single area.

File: code/_helpers/maths.dm (hunk 1 above)
Comment (skull132) [inline code/_helpers/maths.dm]: Debatable.

File: code/game/objects/items/devices/powersink.dm (hunk 2 above)
Comment (skull132) [inline code/game/objects/items/devices/powersink.dm]: ~1200 buckets. If that counts for timers.

**Pull request merged**